oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -560,7 +560,19 @@ impl Runtime {
         node_id: NodeId,
         label: &Label,
     ) -> Result<(oak_abi::Handle, oak_abi::Handle), OakStatus> {
+        info!("creating Channel with label: {:?}", label);
+        if self.is_terminating() {
+            return Err(OakStatus::ErrTerminated);
+        }
+
+        // The label (and mere presence) of the newly created Channel is effectively public, so we
+        // must ensure that the label of the calling Node flows to both "public untrusted" and to
+        // the label of the Channel to be created.
+        self.validate_can_write_to_label(node_id, &Label::public_untrusted())?;
```

Comment:
This should be public_trusted - the bottom of the information flow lattice.

Reply:
I suspect you are missing a way to represent public_trusted ?

Reply:
I am still confused whenever this comes up, which makes me feel I am probably missing something basic.

The current model is more similar to Flume in terms of how labels are composed by set of tags, so there is no way of even representing "fully trusted", right? Or would that be the set of all possible integrity tags (which anyways is not practically representable)?

Reply:
In Flume, fully trusted would be the set of all tags, which is difficult to represent. 

What you could do instead is:
Have a special label which is not a set of tags, but is just called trusted. Then change meet/join/order for the integrity lattice so that:
- meet(x, trusted) = trusted, meet(trusted, x) = trusted
- join(x, trusted) = x, join(trusted, x) = x
- trusted orderedBefore X is always true
- x orderedBefore trusted is never true (unless x is trusted)
For the cases in which "trusted" is not involved you just do the normal set theory thing.

You probably want to do the same trick for secret (the TOP) of the confidentiality lattice (with the direction of everything flipped). So maybe you really want to just call this special symbol "universe".

In case it is helpful:
JoinLabelComp / MeetLabelComp here sort of do something like this:
https://github.com/apl-cornell/sirrtl/blob/master/src/main/scala/firrtl/ir/LabelComp.scala

Reply:
Thanks, I think that would make sense (I'll create a separate issue to track the introduction of the special universe label). Though if the "label of labels" is really "public *trusted*", then it sounds like no individual node should be able to influence that? Since usually they would not be fully trusted. But maybe we go back to your other point that the initial node in an Oak Application should really be *trusted*? Though that still does not sound right to me, why would it be trusted?

Reply:
Also in case it is helpful:

https://github.com/apl-cornell/sirrtl/blob/master/src/main/scala/firrtl/ir/Label.scala

Labels here are pairs with conf/integ, just like in Flume, and then the LabelComps are one part of the pair (so similar to a single set of tags in Oak/Flume). The meet/join for Labels (the pairs) cause meets joins over the components to happen. For Oak, the labels would still be pairs, and the label components would either be sets or this universe symbol.

(This code is not exactly what you want, because in the code meets/joins over labelcomponents don't involve sets. They are more like a logic. To do flowsTo with this, you would need a SAT solver. Maybe it still gives you an idea though).

Reply:
> But maybe we go back to your other point that the initial node in an Oak Application should really be **trusted**? Though that still does not sound right to me, why would it be trusted?

If the first node is the one that decides all the labels, then there really is a trust assumption there. You are trusting this node to get all the labels "right". Someone wanting to decide if an Oak App is secure should really go look at this node and inspect the labels.

Separately, from an IFC perspective, if an "untrusted" thing can decide the labels of anything that isn't also totally untrusted, then this is not allowed.

Reply:
> Someone wanting to decide if an Oak App is secure should really go look at this node and inspect the labels.

Just so we are clear, if this turns out to actually be the case, then we don't have much of a use case for Oak. In particular, I **think** this is not a necessary assumption for us because of the role of the Oak Runtime, which is itself actually trusted, therefore the first Node in an Application can actually be untrusted. But I am struggling to reconcile this with classic IFC literature.

Reply:
The labels are the (low-level) policy. Doesn't the policy have to be trusted ?

Reply:
I think the initial node should by convention:
- create nodes/channels
- assign labels to nodes/channels
- have zero application code in it.

Reply:
What is trusted is the fact that the Oak Runtime will, say, give the integrity of U_0 only to nodes that are actually controlled by U_0 (say, operated by the user U_0 over gRPC). In particular, the initial node of an application cannot claim this (or any other) integrity for itself, or for any node it creates. So maybe the solution is just to go one level up, and say the Oak Runtime is really "public trusted", and then it creates gRPC pseudo-nodes which are less trusted than itself, but still maintain the integrity of the actual user connected on the other side, and also it creates the initial node of the application as "public untrusted"? Note that after this PR, this node is able to create other nodes, but because it is itself untrusted, those nodes can also never be themselves trusted. Though some of them (e.g. a Wasm node created later on) may have the privilege to endorse data (with an integrity label corresponding to its own hash, for instance), so although its own integrity label is still untrusted, it can actually write to (some) more trusted locations.

Reply:
I think having the runtime decide the integrity labels of individual nodes has the effect of baking application-specific behavior into the runtime. The policy for an application is application specific, so it should be the input to the application. 

Maybe it makes sense to think of the node for assigning labels as the "zeroith" node, or the policy node. The first node that runs application code can then be untrusted if this is the policy.

Maybe integrity labels are just confusing and not really used and we should just not have them (for now, or indefinitely)? 
> Note that after this PR, this node is able to create other nodes, but because it is itself untrusted ... some of them (e.g. a Wasm node created later on) may have the privilege to endorse data

I think this PR has the effect of completely turning off integrity labels (which may be fine). Only untrusted nodes can be created, so there is no need to endorse data, because there are no nodes that are more trusted than "untrusted".

Reply:
I guess I will just approve so progress can be made, anyway.

Reply:
But backing up a bit, I think ideally what we want is:

- The policy is something that is trusted. This is a description of "good behavior" for the application
- The application code is untrusted. We want to use _something_ to show that the untrusted code actually _does_ obey this trusted policy.

In doing so, the code is synthetically shown to be trustworthy.

If we are trying to do that with IFC, I think "the policy" is the labels, and so the labels (+ mapping of labels to code) needs to be trusted.

I think the particular policy is also application-dependent, so this needs to be an input to the application.

So I think we need a trusted input for each application that describes the labels. I think the zeroith node could be a way to do this. Maybe some kind of config file is another way?

Reply:
Although, "low-level policy" Is probably both of:
- labels assigned to nodes/channels
- **the declassifications / endorsements**

And the downgrades are ABI calls that happen at runtime and can't as easily go into a config file / single initial node.

Reply:
Can we avoid the term "policy"? I feel it is already overloaded as it is. :slightly_smiling_face: 

As a user, I want to send my data to a random (untrusted) Oak Application, running on a legitimate (trusted) Oak Runtime. I can label my data with a confidentiality label (potentially referring to other users, or to individual wasm nodes), and I want to know that, no matter what the untrusted application does, my data will not be revealed to unwanted parties. I am fine with random nodes processing my data, as long as the confidentiality label is kept. Basically what I want is that I expect the Oak Runtime to assign declassification privilege correctly, rather than the Oak Application assigning labels correctly. But I am aware that at some point those concepts actually overlap, so perhaps I am just holding it wrong.

Reply:
> Basically what I want is that I expect the Oak Runtime to assign declassification privilege correctly, rather than the Oak Application assigning labels correctly.

Don't both of those things depend on what the application is? And then, don't we need some way of telling the oak runtime how to do that for a particular application?

Reply:
> If we are trying to do that with IFC, I think "the policy" is the labels, and so the labels (+ mapping of labels to code) needs to be trusted.

I would have said that the **privilege** to declassify labels is the actual "policy". 

i.e. I am fine with a random untrusted node having a confidentiality label of "user:tzn"; anyone should be able to create a new node with arbitrary code and label it with that (so that, say, it operates on my data, and its output is also equally labelled), as long as it does not have the privilege to **declassify** that label.

Reply:
**This Commit**
I am currently thinking changing the rule for channel/node creation so that it can only be done by some constant and public label might not be best.

I think the only potential leaks that this change prevents are those through the use of introspection. I think introspection makes the portion of the control-flow graph that crosses security levels public. I think this might limit expressiveness significantly. Can we instead consider making introspection a debug-mode-only feature? To do this, a config including the debug mode option would need to be covered by the remote attestation measurement.  What is the use of introspection?

I think if we imagine the runtime in which declassification is disabled, and we do not have introspection, and we use the old rule where we check that the label of the caller node flowsTo the label of the node/channel being created, this "is sound".

It might also look like this fixes "Encoding Information Flow in the Data Flow Path", but I think this problem is also about what we can prove when declassification _is_ involved. I think this commit is not going to fix this problem. (I can expand on this if requested).

**The Term Policy**
I think without using the term policy, I have a hard time describing what we are trying to use IFC to do. And if we don't have a good way of describing the goals of the system, I think we'll have a hard time deciding if the system meets those goals and what the assumptions are. I think the word policies is one way of capturing what we are trying to use IFC to do, but maybe:
- it is not the only way of describing policies that we want to have, ultimately. (We want something else that is more intuitive to humans like users, experts, programmers ?)
- we don't think IFC alone is enough, (so really it just enforces some subset of the policy?)

I would be fine to use other terms, but I don't know what to use. What if we have "high-level policy" as something that humans read and write, and "low-level policy" as something that is enforced by IFC ? What other term should I use (this is not a rhetorical question :) ) ?

Or what about "security goal" ? "Low-level security goal" ?

**Are Labels Trusted**
When designing secure systems even separately from IFC, to consider something untrusted, it should be possible to both: 1) assume it has arbitrarily bad behavior, and 2) still achieve your security goal. 

To claim that labels are not trusted, we need the system to use arbitrary labels and then still achieve the goal.

Assume labels are untrusted. If the human Tiziano sends a piece of data to a remote system, but he does not know the labels that are used in this remote system, the system could use the label "public" for his data. The IFC rules would allow this data to be sent anywhere. I believe both labels and downgrades need to be trusted.

(The ipad scribble about purple labels might help weaken this trust assumption somewhat).

**Representing the Universe**
(I am amused by this header)

Actually, Gogul said kind of the same thing about using a special symbol for universe and also referred to this code: https://github.com/PolymerLabs/arcs/blob/master/java/arcs/core/analysis/BoundedAbstractElement.kt

So this is also possibly helpful. (I forgot this previously, sorry!)

Reply:
> Representing the Universe
> (I am amused by this header)

**To infinity and beyond!**

(presumably via the upward Löwenheim–Skolem theorem&hellip;)

Reply:
> I think if we imagine the runtime in which declassification is disabled, and we do not have introspection, and we use the old rule where we check that the label of the caller node flowsTo the label of the node/channel being created, this "is sound".

But I think when we reintroduce declassification (which presumably is needed for any "interesting" example), then it will bite us back again? In particular, we assume that nodes can inspect their own label, and the label of channels they are passed in, so we are already effectively treating labels as public, and so I think it makes sense to actually enforce this. The alternative would be for labels to always have "top" confidentiality label? Or for a label to have itself as label?

Reply:
|But I think when we reintroduce declassification (which presumably is needed for any "interesting" example), then it will bite us back again?

I don't think this commit will help us get stronger security when there is downgrading. With downgrades, the strongest result we can get is roughly:
- There is a node X that does downgrading to level L
- We form a graph in which channel sends are edges, nodes are nodes
- All information contained in nodes that reach X at this graph is visible at level L (even if it wouldn't normally be in the lattice order).

(Robust declassification could in principal strengthen this a bit by requiring node X to have a minimum integrity label, but still what actually gets released if it can do the downgrade is the above).

Making channel create / node create fixed to a constant label won't help us get a stronger result.

Reply:
For a node at label L to make a new node at label L'

- checking that L flowsTo untrusted makes no sense. This means only nodes that nobody trusts at all can make new nodes.
- checking that L flowsTo trusted makes sense. Everyone trusts trusted. It might be overly restrictive
- checking that L flowsTo L' makes sense. It means that L' trusts L. For example, you could have a trusted router that spawns a node in response to a request by Tiziano and makes a node with integrity user:tzn, where trusted flowsTo user:tzn)
- checking that L flowsTo L' and L flowsTo untrusted is technically allowed, but is awkward. Downgrades are meant to be the exception, not the normal case. Another way to look at it: If we were to switch to a system in which we could prove NMIFC, we'd use integrity to do declassification and confidentiality to do endorsement. If the label of the first node in the system is "public untrusted" then this node would not be able to downgrade anything. So every node in the system would have to be "public untrusted" as well which means you are not using IFC.

Reply:
How about this: 

- the Oak Runtime "public trusted" (as a side node, we don't even need to explicitly represent its label in this model)
- the main WebAssembly node of an Oak Application is "public untrusted"
- the (untrusted) WebAssembly node can **request** the (trusted) Runtime to create any additional nodes and channels
- the Runtime actually checks something about each such request, and then eventually decides whether to create the requested node or not, and at which label. For instance, if the calling node is itself untrusted, the Runtime will refuse to create a new node with a more trusted label.

Is this still problematic?

Reply:
What you proposed is fine if "The Oak Runtime" really is a node and really is labeled public trusted. This real node should be the one making the create_node and create_channel calls, then. In this case, what you described would be permitted by either of the following two options:
- checking that L flowsTo public_trusted 
- checking that L flowsTo L' 

If instead, the public_untrusted WebAssembly node is the one really doing the channel_create and node_create calls, then this is problematic because then we are in the 4th bullet point above.

Reply:
Let's take a step back.

If we only had confidentiality labels, then I think we agree that the check simplifies to L flows_to public (or, better, L can_declassify_to public ?). This represents the fact that labels are really public entities (let's stick with this assumption for now, even though we considered alternatives), and creating a node or channel really exposes that information to the public.

IIUC, the introduction of integrity labels complicates things, because under IFC rules, a node with high integrity would not be able to read a low integrity value, and therefore if labels were untrusted, no trusted node would be able to actually read them. Is this a correct intuition @aferr ? Is there a better way of explaining this?

But now this implies that the creator of the node / channel needs to have label L flows_to trusted (for the integrity component), which is true only for a node that is already fully trusted. Though requiring this seems to me excessive, because IIUC for a node to be fully trusted it means that it may affect nodes / channels with arbitrary integrity (which seems a very powerful privilege in itself).

In our case, I think the assumption of any one application node being fully trusted seems too strong. In particular, I don't see why an untrusted node should not be allowed to create other nodes; I trust your point that there is a reason, but my intuition about integrity labels is not strong enough for me to see it.

@aferr do you have a simple example of something that would go wrong if, say, the Oak Runtime transparently endorsed labels to fully trusted, even if the operation is performed by an untrusted node?

Also AFAICT Flume does not have this problem, is it just skipping over it, or is it solved via some other mechanism?

Apologies to keep dragging this, I am actually learning a lot from the discussion, and even repeating similar points in slightly different ways helps me (and I hope others) build a more solid and intuitive understanding of such concepts, so I am really grateful for the time and effort you put into this @aferr .

Reply:
The explanation you gave as to why integrity has to go from more trusted to less or equal integrity is right.

One more way of looking at it: 
If node A has integrity X, then the code in A is trusted about X. If node B is creating node A, then B is really supplying the code in A. So B should also be trusted about X. 

So you are right, if we want to make the label constant, the check would have to be L(creator) flowsTo public_trusted.

You are also right that "trusted" is quite powerful. So I understand why you want to limit its use. I _think_ in practice, many applications will probably have some components that do need to be universally trusted, but we will want to keep it a goal to limit this as much as possible. As a starting point, if the initial node is public_trusted, we can stick to the convention above of trying to just use this node to create other nodes/channels. (We can also think of this node as giving up privilege.)

I think the check L(creator) flowsTo public_trusted is safe, but is actually overly restrictive if we remove introspection. Introspection seems like it will limit what we can do a lot because it makes the control flow graph public. This means no implicit flows are prevented, and in general implicit flows can contain anything. (It's very easy to create a program that leaks an arbitrary secret through control flow). The check L(creator) flowsTo L(created) should give the same security, even when downgrades are permitted. Do you buy this, or should I expand?

Re transparent endorsement:
If we use robust declassification / transparent endorsement / non-malleable IFC,
public_untrusted actually cannot downgrade anything. It can't declassify because it's untrusted and it can't endorse because it's public.

Re Flume:
It sounds like Flume does have the same restriction when creating objects (L(creator) flowsTo L(created))
Page 4, Section 3.3 under the header Objects
"When a process p creates an object o, p specifies o’s labels, subject to the restriction that p must be able to write to o"
Checking that p can write to o is the flowsTo check, given in Definition 3 (it checks that p is less secret and more trusted than o).

Reply:
I have created https://github.com/project-oak/oak/issues/1213 to keep discussing this, but I'd like to merge this PR in the meanwhile. Happy to go back and change things in the code when we do arrive to a conclusion, I just think an issue is probably better than a long thread on this PR.